src/NuGet.Core/NuGet.Protocol.Core.v3/PackageUpdateResource.cs

```diff
@@ -0,0 +1,601 @@
+            }
+            else
+            {
+                await PushPackageToServer(source, apiKey, packageToPush, new FileInfo(packageToPush).Length, log, token);
+            }
+
+            log.LogInformation(string.Format(CultureInfo.CurrentCulture, Strings.PushCommandPackagePushed));
+        }
+
+        private static string GetSourceDisplayName(string source)
+        {
+            if (String.IsNullOrEmpty(source) || source.Equals(NuGetConstants.DefaultGalleryServerUrl, StringComparison.OrdinalIgnoreCase))
+            {
+                return Strings.LiveFeed + " (" + NuGetConstants.DefaultGalleryServerUrl + ")";
+            }
+            if (source.Equals(NuGetConstants.DefaultSymbolServerUrl, StringComparison.OrdinalIgnoreCase))
+            {
+                return Strings.DefaultSymbolServer + " (" + NuGetConstants.DefaultSymbolServerUrl + ")";
+            }
+            return "'" + source + "'";
+        }
+
+        private static IEnumerable<string> GetPackagesToPush(string packagePath)
+        {
+            // Ensure packagePath ends with *.nupkg
+            packagePath = EnsurePackageExtension(packagePath);
+            return PerformWildcardSearch(Directory.GetCurrentDirectory(), packagePath);
+        }
+
+        private static string EnsurePackageExtension(string packagePath)
+        {
+            if (packagePath.IndexOf('*') == -1)
+            {
+                // If there's no wildcard in the path to begin with, assume that it's an absolute path.
+                return packagePath;
+            }
+            // If the path does not contain wildcards, we need to add *.nupkg to it.
+            if (!packagePath.EndsWith(NuGetConstants.PackageExtension, StringComparison.OrdinalIgnoreCase))
+            {
+                if (packagePath.EndsWith("**", StringComparison.OrdinalIgnoreCase))
+                {
+                    packagePath = packagePath + Path.DirectorySeparatorChar + '*';
+                }
+                else if (!packagePath.EndsWith("*", StringComparison.OrdinalIgnoreCase))
+                {
+                    packagePath = packagePath + '*';
+                }
+                packagePath = packagePath + NuGetConstants.PackageExtension;
+            }
+            return packagePath;
+        }
+
+        private static void EnsurePackageFileExists(string packagePath, IEnumerable<string> packagesToPush)
+        {
+            if (!packagesToPush.Any())
+            {
+                throw new ArgumentException(String.Format(CultureInfo.CurrentCulture,
+                    Strings.UnableToFindFile,
+                    packagePath));
+            }
+        }
+
+        // Indicates whether the specified source is a file source, such as: \\a\b, c:\temp, etc.
+        private static bool IsFileSource(string source)
+        {
+            Uri uri;
+            if (Uri.TryCreate(source, UriKind.RelativeOrAbsolute, out uri))
+            {
+                return uri.IsFile;
```

Comment:
Verify that this works with relative source paths like "nupkgs"  if not you might be better off doing !uri.IsHttp